pkg/extensions/tbtc/tbtc.go

```diff
@@ -0,0 +1,286 @@
+//  2. Incoming events deduplication.
+//  3. Resume monitoring after client restart.
+func (t *tbtc) monitorAndAct(
+	ctx context.Context,
+	monitoringName string,
+	monitoringStartFn watchDepositEventFn,
+	monitoringStopFn watchDepositEventFn,
+	keepClosedFn watchKeepClosedFn,
+	actFn submitDepositTxFn,
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) (subscription.EventSubscription, error) {
+	handleStartEvent := func(deposit string) {
+		logger.Infof(
+			"starting [%v] monitoring for deposit [%v]",
+			monitoringName,
+			deposit,
+		)
+
+		stopEventChan := make(chan struct{})
+
+		stopEventSubscription, err := monitoringStopFn(
+			func(stopEventDeposit string) {
+				if deposit == stopEventDeposit {
+					stopEventChan <- struct{}{}
+				}
+			},
+		)
+		if err != nil {
+			logger.Errorf(
+				"could not setup stop event handler for [%v] "+
+					"monitoring for deposit [%v]: [%v]",
+				monitoringName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer stopEventSubscription.Unsubscribe()
+
+		keepClosedChan, keepClosedUnsubscribe, err := keepClosedFn(deposit)
+		if err != nil {
+			logger.Errorf(
+				"could not setup keep closed handler for [%v] "+
+					"monitoring for deposit [%v]: [%v]",
+				monitoringName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer keepClosedUnsubscribe()
+
+		timeoutChan := time.After(timeout)
+
+		actionAttempt := 1
+
+	monitoring:
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Infof(
+					"context is done for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-stopEventChan:
+				logger.Infof(
+					"stop event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-keepClosedChan:
+				logger.Infof(
+					"keep closed event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-timeoutChan:
+				err := actFn(deposit)
+				if err != nil {
+					if actionAttempt == maxActionAttempts {
+						logger.Errorf(
+							"could not perform action "+
+								"for [%v] monitoring for deposit [%v]: [%v]; "+
+								"last attempt failed",
+							monitoringName,
+							deposit,
+							err,
+						)
+						break monitoring
+					}
+
+					backoff := actBackoffFn(actionAttempt)
+
+					logger.Errorf(
+						"could not perform action "+
+							"for [%v] monitoring for deposit [%v]: [%v]; "+
+							"retrying after: [%v]",
+						monitoringName,
+						deposit,
+						err,
+						backoff,
+					)
+
+					timeoutChan = time.After(backoff)
+					actionAttempt++
+				} else {
+					logger.Infof(
+						"action for [%v] monitoring for "+
+							"deposit [%v] has been performed successfully",
+						monitoringName,
+						deposit,
+					)
+					break monitoring
+				}
+			}
+		}
+
+		logger.Infof(
+			"[%v] monitoring for deposit [%v] "+
+				"has been completed",
```

Comment:
I am not sure about `has been completed`, IMO it suggests it's a happy exit while we could reach the maximum number of retries. Let's maybe say `stopped [%v] monitoring for deposit [%v]`. WDYT?